immutable/src/main/java/org/jdbi/v3/core/mapper/reflect/ImmutableMapper.java

```diff
@@ -0,0 +1,308 @@
+                                return new ArrayList<>(); // TODO IMMUTABLE LIST
+                            }
+                            if (returnType.equals(Set.class)) {
+                                return new HashSet<>(); // TODO IMMUTABLE SET
+                            }
+                            throw new UnsupportedOperationException(method.getName());
+                        }
+                    }
+                    return invocationMap.get(method);
+                }
+            });
+
+            if (!unmatchedDescriptors.isEmpty()) {
+                for (PropertyDescriptor descriptor : unmatchedDescriptors) {
+                    final Method method = descriptor.getReadMethod();
+                    if (!method.isDefault()) {
+                        Class<?> returnType = method.getReturnType();
+                        if (!returnType.equals(Map.class) && !returnType.equals(List.class) && !returnType.equals(Set.class)) {
+                            throw new MappingException("Found Method without default implementation in Immutable");
+                        }
+                    }
+                }
+            }
+
+            return (T) proxy;
+        };
+    }
+
+    private static String paramName(PropertyDescriptor descriptor) {
+        return Stream.of(descriptor.getReadMethod(), descriptor.getWriteMethod())
+            .filter(Objects::nonNull)
+            .map(method -> method.getAnnotation(ColumnName.class))
+            .filter(Objects::nonNull)
+            .map(ColumnName::value)
+            .findFirst()
+            .orElseGet(descriptor::getName);
+    }
+
+    private String debugName(PropertyDescriptor descriptor) {
+        return String.format("%s.%s", type.getSimpleName(), descriptor.getName());
+    }
+
+    private List<PropertyDescriptor> getInformation(Class<T> immutable) {
+        final List<Method> getter = Arrays.asList(immutable.getMethods());
+
+        return getter.stream().filter(g -> g.getParameterCount() == 0).map(g -> {
+            String rawName = getRawName(g.getName());
+            try {
+                return new PropertyDescriptor(rawName,
+                    g,
+                    null);
+            } catch (IntrospectionException e) {
+                throw new MappingException(e);
+            }
+        }).collect(Collectors.toList());
+    }
+
+    private String getRawName(String name) {
+        if (name.startsWith("get")) {
+            if (Character.isUpperCase(name.charAt(3))) {
+                return Character.toLowerCase(name.charAt(3)) + name.substring(4);
+            }
+        }
+        if (name.startsWith("is")) {
+            if (Character.isUpperCase(name.charAt(2))) {
+                return Character.toLowerCase(name.charAt(2)) + name.substring(3);
+            }
+        }
+        return name;
+    }
+
+    private Object invokeDefault(Method method, Object proxy, Class<?> declaringClass, Object[] args) throws Throwable {
```

Comment:
We already have similar code in SQL object (DefaultMethodHandler).

Let's extract the relevant code from there into an internal utility class in core, and reuse it from here rather than reimplementing.